src/Raven.Server/RavenServer.cs

```diff
@@ -929,45 +938,179 @@ internal AuthenticateConnection AuthenticateConnectionCertificate(X509Certificat
+                            // If connection will be approved, cert won't be null anymore and will be assigned the relevant permissions
+                            MaybeAllowConnectionBasedOnPinningHash(certificate, ctx, ref authenticationStatus, ref cert, remoteAddress);
                         }
-                        else
+
+                        if (cert != null)
                         {
-                            authenticationStatus.Status = AuthenticationStatus.Allowed;
-                            foreach (var kvp in definition.Permissions)
+                            var definition = JsonDeserializationServer.CertificateDefinition(cert);
+                            authenticationStatus.Definition = definition;
+                            if (definition.SecurityClearance == SecurityClearance.ClusterAdmin)
                             {
-                                authenticationStatus.AuthorizedDatabases.Add(kvp.Key, kvp.Value);
+                                authenticationStatus.Status = AuthenticationStatus.ClusterAdmin;
+                            }
+                            else if (definition.SecurityClearance == SecurityClearance.ClusterNode)
+                            {
+                                authenticationStatus.Status = AuthenticationStatus.ClusterAdmin;
+                            }
+                            else if (definition.SecurityClearance == SecurityClearance.Operator)
+                            {
+                                authenticationStatus.Status = AuthenticationStatus.Operator;
+                            }
+                            else
+                            {
+                                authenticationStatus.Status = AuthenticationStatus.Allowed;
+                                foreach (var kvp in definition.Permissions)
+                                {
+                                    authenticationStatus.AuthorizedDatabases.Add(kvp.Key, kvp.Value);
+                                }
                             }
                         }
                     }
+                    finally
+                    {
+                        ctx.CloseTransaction();
+                    }
                 }
             }
 
             return authenticationStatus;
         }
 
+        private void MaybeAllowConnectionBasedOnPinningHash(X509Certificate2 certificate, TransactionOperationContext ctx,
+            ref AuthenticateConnection authenticationStatus, ref BlittableJsonReaderObject cert, string remoteAddress)
+        {
+            // The certificate is not explicitly registered in our server, let's see if we have a certificate
+            // with the same public key pinning hash.
+            var pinningHash = certificate.GetPublicKeyPinningHash();
+            var certsWithSameHash = ServerStore.Cluster.GetCertificatesByPinningHashSortedByExpiration(ctx, pinningHash);
+
+            if (certsWithSameHash.Count > 0)
```

Comment:
Let's reverse this if statement and reduce nesting in this way

Reply:
OK